pkg/credentials/file_store.go

```diff
@@ -0,0 +1,82 @@
+/*
+Copyright the Velero contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package credentials
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	kbclient "sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/vmware-tanzu/velero/pkg/util/kube"
+
+	"github.com/vmware-tanzu/velero/pkg/util/filesystem"
+)
+
+// FileStore defines operations for interacting with credentials
+// that are stored on a file system.
+type FileStore interface {
+	// Get returns a path on disk where the secret defined by the given
+	// selector is serialized.
+	Get(selector *corev1.SecretKeySelector) (string, error)
+}
+
+type namespacedFileStore struct {
+	client    kbclient.Client
+	namespace string
+
+	fsRoot string
+	fs     filesystem.Interface
+}
+
+func (n *namespacedFileStore) Get(selector *corev1.SecretKeySelector) (string, error) {
+	creds, err := kube.GetSecretKey(n.client, n.namespace, selector)
```

Comment:
Would it make sense to look for an existing file on disk before fetching the key? Or do we always fetch in order to ensure that the key doesn't get stale in case of updates?

Reply:
I thought about doing that but opted to always fetch to avoid the credentials being stale. Thankfully we're using the kubebuilder client here so there is caching built in to avoid spamming the API server.

Reply:
My intention was that we could use this approach for now but could always optimise it in future. I experimented with creating a controller that would watch for and serialize any secrets from the velero namespace and update them upon change but I didn't think that I could guarantee that the latest version of a secret would always be available. It seemed safest to always refresh it before use.